# Guided Project: Exploring Ebay Car Sales Data

In this guided project, we'll work with a dataset of used cars from eBay Kleinanzeigen, a classifieds section of the German eBay website. The dataset was originally scraped and uploaded to Kaggle. A few modifications from the original dataset that was uploaded to Kaggle have been made:

50,000 data points from the full dataset have been sampled, to ensure your code runs quickly in our hosted environment
The dataset has been dirtied a bit to more closely resemble what you would expect from a scraped dataset (the version uploaded to Kaggle was cleaned to be easier to work with).
The aim of this project is to clean the data and analyze the included used car listings.

In [1]:
import numpy as np
import matplotlib.pyplot as plt
import pandas as pd
%matplotlib inline

In [2]:
df = pd.read_csv("/storage/emulated/0/Documents/projects/Ebay/autos.csv",encoding='Latin-1')

In [3]:
df.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 50000 entries, 0 to 49999
Data columns (total 20 columns):
 #   Column               Non-Null Count  Dtype 
---  ------               --------------  ----- 
 0   dateCrawled          50000 non-null  object
 1   name                 50000 non-null  object
 2   seller               50000 non-null  object
 3   offerType            50000 non-null  object
 4   price                50000 non-null  object
 5   abtest               50000 non-null  object
 6   vehicleType          44905 non-null  object
 7   yearOfRegistration   50000 non-null  int64 
 8   gearbox              47320 non-null  object
 9   powerPS              50000 non-null  int64 
 10  model                47242 non-null  object
 11  odometer             50000 non-null  object
 12  monthOfRegistration  50000 non-null  int64 
 13  fuelType             45518 non-null  object
 14  brand                50000 non-null  object
 15  notRepairedDamage    40171 non-null  object
 16  date

In [4]:
df.columns

Index(['dateCrawled', 'name', 'seller', 'offerType', 'price', 'abtest',
       'vehicleType', 'yearOfRegistration', 'gearbox', 'powerPS', 'model',
       'odometer', 'monthOfRegistration', 'fuelType', 'brand',
       'notRepairedDamage', 'dateCreated', 'nrOfPictures', 'postalCode',
       'lastSeen'],
      dtype='object')

In [5]:
new_columns = {'dateCrawled':'date_crawled', 'name':'name', 'seller':'seller', 
                'offerType':'offer_type', 'price':'price', 'abtest':'ab_test',
               'vehicleType':'vehicle_type', 'yearOfRegistration':'registration_year', 
               'gearbox':'gearbox','powerPS':'power_ps', 'model':'model','odometer':'odometer', 
               'monthOfRegistration':'registration_month', 'fuelType':'fuel_type', 'brand':'brand',
             'notRepairedDamage':'unrepaired_damage', 'dateCreated':'date_created',
              'nrOfPictures':'no_of_pictures', 'postalCode':'postal_code','lastSeen':'last_seen'
    }

In [6]:
df.rename(columns=new_columns, inplace= True)

In [7]:
df.head()

date_crawled                                               name  \
0  2016-03-26 17:47:46                   Peugeot_807_160_NAVTECH_ON_BOARD   
1  2016-04-04 13:38:56         BMW_740i_4_4_Liter_HAMANN_UMBAU_Mega_Optik   
2  2016-03-26 18:57:24                         Volkswagen_Golf_1.6_United   
3  2016-03-12 16:58:10  Smart_smart_fortwo_coupe_softouch/F1/Klima/Pan...   
4  2016-04-01 14:38:50  Ford_Focus_1_6_Benzin_TÜV_neu_ist_sehr_gepfleg...   

   seller offer_type   price  ab_test vehicle_type  registration_year  \
0  privat    Angebot  $5,000  control          bus               2004   
1  privat    Angebot  $8,500  control    limousine               1997   
2  privat    Angebot  $8,990     test    limousine               2009   
3  privat    Angebot  $4,350  control   kleinwagen               2007   
4  privat    Angebot  $1,350     test        kombi               2003   

     gearbox  power_ps   model   odometer  registration_month fuel_type  \
0    manuell       158  andere  150,000km                   3       lpg   
1  automatik       286     7er  150,000km                   6    benzin   
2    manuell       102    golf   70,000km                   7    benzin   
3  automatik        71  fortwo   70,000km                   6    benzin   
4    manuell         0   focus  150,000km                   7    benzin   

        brand unrepaired_damage         date_created  no_of_pictures  \
0     peugeot              nein  2016-03-26 00:00:00               0   
1         bmw              nein  2016-04-04 00:00:00               0   
2  volkswagen              nein  2016-03-26 00:00:00               0   
3       smart              nein  2016-03-12 00:00:00               0   
4        ford              nein  2016-04-01 00:00:00               0   

   postal_code            last_seen  
0        79588  2016-04-06 06:45:54  
1        71034  2016-04-06 14:45:08  
2        35394  2016-04-06 20:15:37  
3        33729  2016-03-15 03:16:28  
4        39218  2016-04-01 14:38:50

In [8]:
df.describe(include='all')

date_crawled         name  seller offer_type  price ab_test  \
count                 50000        50000   50000      50000  50000   50000   
unique                48213        38754       2          2   2357       2   
top     2016-03-22 09:51:06  Ford_Fiesta  privat    Angebot     $0    test   
freq                      3           78   49999      49999   1421   25756   
mean                    NaN          NaN     NaN        NaN    NaN     NaN   
std                     NaN          NaN     NaN        NaN    NaN     NaN   
min                     NaN          NaN     NaN        NaN    NaN     NaN   
25%                     NaN          NaN     NaN        NaN    NaN     NaN   
50%                     NaN          NaN     NaN        NaN    NaN     NaN   
75%                     NaN          NaN     NaN        NaN    NaN     NaN   
max                     NaN          NaN     NaN        NaN    NaN     NaN   

       vehicle_type  registration_year  gearbox      power_ps  model  \
count         44905       50000.000000    47320  50000.000000  47242   
unique            8                NaN        2           NaN    245   
top       limousine                NaN  manuell           NaN   golf   
freq          12859                NaN    36993           NaN   4024   
mean            NaN        2005.073280      NaN    116.355920    NaN   
std             NaN         105.712813      NaN    209.216627    NaN   
min             NaN        1000.000000      NaN      0.000000    NaN   
25%             NaN        1999.000000      NaN     70.000000    NaN   
50%             NaN        2003.000000      NaN    105.000000    NaN   
75%             NaN        2008.000000      NaN    150.000000    NaN   
max             NaN        9999.000000      NaN  17700.000000    NaN   

         odometer  registration_month fuel_type       brand unrepaired_damage  \
count       50000        50000.000000     45518       50000             40171   
unique         13                 NaN         7          40                 2   
top     150,000km                 NaN    benzin  volkswagen              nein   
freq        32424                 NaN     30107       10687             35232   
mean          NaN            5.723360       NaN         NaN               NaN   
std           NaN            3.711984       NaN         NaN               NaN   
min           NaN            0.000000       NaN         NaN               NaN   
25%           NaN            3.000000       NaN         NaN               NaN   
50%           NaN            6.000000       NaN         NaN               NaN   
75%           NaN            9.000000       NaN         NaN               NaN   
max           NaN           12.000000       NaN         NaN               NaN   

               date_created  no_of_pictures   postal_code            last_seen  
count                 50000         50000.0  50000.000000                50000  
unique                   76             NaN           NaN                39481  
top     2016-04-03 00:00:00             NaN           NaN  2016-04-07 06:17:27  
freq                   1946             NaN           NaN                    8  
mean                    NaN             0.0  50813.627300                  NaN  
std                     NaN             0.0  25779.747957                  NaN  
min                     NaN             0.0   1067.000000                  NaN  
25%                     NaN             0.0  30451.000000                  NaN  
50%                     NaN             0.0  49577.000000                  NaN  
75%                     NaN             0.0  71540.000000                  NaN  
max                     NaN             0.0  99998.000000                  NaN

In [9]:
df.price.unique()

array(['$5,000', '$8,500', '$8,990', ..., '$385', '$22,200', '$16,995'],
      dtype=object)

In [10]:
df['price'] = df.price.str.replace(',','').str.replace('$','').astype(int)

In [11]:
df['price'].describe()

count    5.000000e+04
mean     9.840044e+03
std      4.811044e+05
min      0.000000e+00
25%      1.100000e+03
50%      2.950000e+03
75%      7.200000e+03
max      1.000000e+08
Name: price, dtype: float64

In [12]:
df['price'].value_counts().sort_index(ascending=False).head(17)

99999999    1
27322222    1
12345678    3
11111111    2
10000000    1
3890000     1
1300000     1
1234566     1
999999      2
999990      1
350000      1
345000      1
299000      1
295000      1
265000      1
259000      1
250000      1
Name: price, dtype: int64

In [13]:
df['price'].value_counts().sort_index(ascending=True).head(10)

0     1421
1      156
2        3
3        1
5        2
8        1
9        1
10       7
11       2
12       3
Name: price, dtype: int64

Looking at the highest prices - it seems there's a huge jump from 350,000 to 999,990. In fact, all prices above 350,000 look as if they are incorrect. Let's drop those rows. Additionally, we see that there are 1421 cars priced at 0 USD. Let's drop these as well

In [14]:
df = df[(df['price']>100)&(df['price']<999990)]
df.price.describe()

count     48090.000000
mean       5946.617426
std        9085.786730
min         110.000000
25%        1250.000000
50%        3099.000000
75%        7500.000000
max      350000.000000
Name: price, dtype: float64

In [15]:
df['odometer'].unique()

array(['150,000km', '70,000km', '50,000km', '80,000km', '10,000km',
       '30,000km', '125,000km', '90,000km', '20,000km', '60,000km',
       '5,000km', '100,000km', '40,000km'], dtype=object)

In [16]:
df['odometer'] = df['odometer'].str.replace('km','').str.replace(',','').astype(float)

In [17]:
df.rename({'odometer':'odometer_km'},axis=1,inplace=True)

In [18]:
df.odometer_km.describe()

count     48090.000000
mean     125921.501352
std       39519.217705
min        5000.000000
25%      125000.000000
50%      150000.000000
75%      150000.000000
max      150000.000000
Name: odometer_km, dtype: float64

In [19]:
df['odometer_km'].value_counts().sort_index(ascending=False).head(10)

150000.0    31117
125000.0     5026
100000.0     2097
90000.0      1731
80000.0      1410
70000.0      1212
60000.0      1151
50000.0      1009
40000.0       814
30000.0       776
Name: odometer_km, dtype: int64

In [20]:
df['odometer_km'].value_counts().sort_index(ascending=True).head(10)

5000.0      747
10000.0     245
20000.0     755
30000.0     776
40000.0     814
50000.0    1009
60000.0    1151
70000.0    1212
80000.0    1410
90000.0    1731
Name: odometer_km, dtype: int64

The odometer values seem rounded which might mean that the sellers had to choose from a list and not enter the actual numbers of the cars.

# Cleaning the date columns¶
There are 5 columns that should represent date values. Some of these columns were created by the crawler, some came from the website itself. Right now, the date_crawled, last_seen, and ad_created columns are all identified as string values by pandas. Because these three columns are represented as strings, we need to convert the data into a numerical representation so we can understand it quantitatively. The other two columns are represented as numeric values.

Let's first understand how the values in the three string columns are formatted. These columns all represent full timestamp values, like so:

In [21]:
df[['date_crawled','date_created','last_seen']][0:5]

date_crawled         date_created            last_seen
0  2016-03-26 17:47:46  2016-03-26 00:00:00  2016-04-06 06:45:54
1  2016-04-04 13:38:56  2016-04-04 00:00:00  2016-04-06 14:45:08
2  2016-03-26 18:57:24  2016-03-26 00:00:00  2016-04-06 20:15:37
3  2016-03-12 16:58:10  2016-03-12 00:00:00  2016-03-15 03:16:28
4  2016-04-01 14:38:50  2016-04-01 00:00:00  2016-04-01 14:38:50

In [22]:
df['date_crawled'].str[:10].value_counts(normalize=True,dropna=False).sort_index()

2016-03-05    0.025390
2016-03-06    0.014057
2016-03-07    0.035953
2016-03-08    0.033229
2016-03-09    0.033001
2016-03-10    0.032356
2016-03-11    0.032647
2016-03-12    0.036952
2016-03-13    0.015721
2016-03-14    0.036640
2016-03-15    0.034227
2016-03-16    0.029445
2016-03-17    0.031503
2016-03-18    0.012872
2016-03-19    0.034727
2016-03-20    0.037763
2016-03-21    0.037284
2016-03-22    0.032917
2016-03-23    0.032294
2016-03-24    0.029445
2016-03-25    0.031399
2016-03-26    0.032231
2016-03-27    0.031129
2016-03-28    0.034997
2016-03-29    0.034082
2016-03-30    0.033791
2016-03-31    0.031836
2016-04-01    0.033687
2016-04-02    0.035600
2016-04-03    0.038615
2016-04-04    0.036556
2016-04-05    0.013100
2016-04-06    0.003161
2016-04-07    0.001393
Name: date_crawled, dtype: float64

The ads were crawled almost daily from on the 5th of march 2016 till 7th of April 2016

In [23]:
df['date_created'].str[:10].value_counts(normalize=True,dropna=False).sort_index()

2015-06-11    0.000021
2015-08-10    0.000021
2015-09-09    0.000021
2015-11-10    0.000021
2015-12-05    0.000021
                ...   
2016-04-03    0.038865
2016-04-04    0.036910
2016-04-05    0.011832
2016-04-06    0.003244
2016-04-07    0.001248
Name: date_created, Length: 76, dtype: float64

We see that the oldest ads were created from 11th june, 2015 and the newest ones - from 7th April, 2016.

In [24]:
df['last_seen'].str[:10].value_counts(normalize=True,dropna=False).sort_index()

2016-03-05    0.001081
2016-03-06    0.004304
2016-03-07    0.005386
2016-03-08    0.007299
2016-03-09    0.009565
2016-03-10    0.010647
2016-03-11    0.012414
2016-03-12    0.023830
2016-03-13    0.008879
2016-03-14    0.012622
2016-03-15    0.015824
2016-03-16    0.016428
2016-03-17    0.028072
2016-03-18    0.007320
2016-03-19    0.015762
2016-03-20    0.020670
2016-03-21    0.020566
2016-03-22    0.021377
2016-03-23    0.018590
2016-03-24    0.019713
2016-03-25    0.019110
2016-03-26    0.016698
2016-03-27    0.015513
2016-03-28    0.020794
2016-03-29    0.022312
2016-03-30    0.024641
2016-03-31    0.023851
2016-04-01    0.022874
2016-04-02    0.024849
2016-04-03    0.025140
2016-04-04    0.024517
2016-04-05    0.125161
2016-04-06    0.221959
2016-04-07    0.132231
Name: last_seen, dtype: float64

The crawler records when it last saw a listing. This allows us to determine when the car was sold.

# Cleaning the registration_year column

In [25]:
df['registration_year'].describe()

count    48090.000000
mean      2004.749927
std         88.018197
min       1000.000000
25%       1999.000000
50%       2004.000000
75%       2008.000000
max       9999.000000
Name: registration_year, dtype: float64

In [26]:
df['registration_year'].value_counts().sort_index()

1000    1
1001    1
1111    1
1800    2
1910    2
       ..
5911    1
6200    1
8888    1
9000    1
9999    3
Name: registration_year, Length: 95, dtype: int64

There are values in the registration_year column that seem incorrect:

Given that cars were invented in 1885, any listing with registration_year prior to that seems to be incorrect.

Given that the ads were created between 11th June, 2015 and 7th April, 2016, any car with registration year after 2016 should also be removed.

Let's count the number of listings with cars that fall outside the 1900 - 2016 interval and see if it's safe to remove those rows entirely, or if we need more custom logic.

In [27]:
(df['registration_year'].shape[0])-(df['registration_year'].between(1900,2016).sum())

1870

We see that 1870 out of the 48090 listings fall outside of the 1900-2016 registration year interval. Given that this account for less than 4%, it is safe to remove them.

In [28]:
df=df[df['registration_year'].between(1900,2016)]

In [29]:
df['registration_year'].value_counts(ascending=False).head(10)

2000    3079
2005    2909
1999    2872
2004    2699
2003    2693
2006    2668
2001    2627
2002    2475
1998    2323
2007    2272
Name: registration_year, dtype: int64

It looks like the majority of the cars were registered between 1999 and 2005.

# EXPLORING THE BRAND COLUMN

In [30]:
df['brand'].value_counts(normalize=True)*100

volkswagen        21.138035
bmw               11.045002
opel              10.703159
mercedes_benz      9.690610
audi               8.693206
ford               6.955863
renault            4.694937
peugeot            2.985720
fiat               2.550844
seat               1.826049
skoda              1.644310
nissan             1.533968
mazda              1.518823
smart              1.423626
citroen            1.404154
toyota             1.282994
hyundai            1.003894
sonstige_autos     0.951969
volvo              0.913025
mini               0.880571
mitsubishi         0.815664
honda              0.785374
kia                0.709650
alfa_romeo         0.668542
porsche            0.603635
suzuki             0.588490
chevrolet          0.566854
chrysler           0.352661
dacia              0.266119
daihatsu           0.248810
jeep               0.229338
land_rover         0.212029
subaru             0.212029
saab               0.166595
jaguar             0.153613
daewoo             0

Unsurprisingly, German brands make up for about 60% of all listings. Volkswagen is the top brand with the number of cars listed from that brand being the same as the total of the next two brands - BMW and Opel. There are a lot of brands that aren't very represented. For the sake of the analysis, we will focus on those that make up more than 5% of the listings.

In [31]:
#Let's isolate the top brands

brand_counts = (df['brand'].value_counts(normalize=True)*100)
top_brands = brand_counts[brand_counts > 5].index

top_brands

Index(['volkswagen', 'bmw', 'opel', 'mercedes_benz', 'audi', 'ford'], dtype='object')

In [32]:
import pprint

In [33]:
# grouping the top brands by their mean price

brand_mean_prices = {}

for brand in top_brands:
    by_brand = df[df['brand']==brand]
    mean_price = by_brand['price'].mean()
    brand_mean_prices[brand] = int(mean_price)
    
pprint.pprint(brand_mean_prices)

{'audi': 9389,
 'bmw': 8384,
 'ford': 3804,
 'mercedes_benz': 8674,
 'opel': 3019,
 'volkswagen': 5452}


Based on the average prices we see that:

Opel and Ford have the lowest average prices.

Audi, Mercedes and BMW have the highest average prices.

Volkswagen are moderately priced.

In [34]:
#grouping the top brands by the mean of their mileage

km_per_brand = {}

for brand in top_brands:
    by_brand = df[df['brand'] == brand]
    mean_km = by_brand['odometer_km'].mean()
    km_per_brand[brand] = int(mean_km)
    
pprint.pprint(km_per_brand)

{'audi': 129260,
 'bmw': 132716,
 'ford': 124248,
 'mercedes_benz': 131053,
 'opel': 129379,
 'volkswagen': 128783}


Above we created a dictionary in which we see the average mileage per brand. Now, let's build a dataframe so that we can easily compare the average price and average mileage per brand and see if we can find any corelation.

In [35]:
mean_prices = pd.Series(brand_mean_prices)
mean_mileage = pd.Series(km_per_brand)

In [36]:
brand_info = pd.DataFrame(mean_prices, columns=['mean_prices'])
brand_info['mean_mileage'] = mean_mileage

brand_info

mean_prices  mean_mileage
volkswagen            5452        128783
bmw                   8384        132716
opel                  3019        129379
mercedes_benz         8674        131053
audi                  9389        129260
ford                  3804        124248

It is interesting to see that even though the average mileage of BMW and Mercedes cars is higher, their prices are still higher than the prices of other brands.

Let's identify columns which have data entered in German and translate it in English. We already know that the 'seller' and 'offer_type' columns have words in German so let's start with them.

In [37]:
df['seller'].unique()

array(['privat'], dtype=object)

In [38]:
df['seller'] = df['seller'].str.replace('privat','private')
df.seller.value_counts()

private    46220
Name: seller, dtype: int64

In [39]:
df['offer_type'].unique()

array(['Angebot'], dtype=object)

In [40]:
df['offer_type'] = df['offer_type'].str.replace('Angebot','offer')
df['offer_type'].value_counts()

offer    46220
Name: offer_type, dtype: int64

In [41]:
df.vehicle_type.value_counts()

limousine     12546
kleinwagen    10439
kombi          8896
bus            4022
cabrio         3004
coupe          2453
suv            1961
andere          386
Name: vehicle_type, dtype: int64

In [42]:
df['vehicle_type'] = df['vehicle_type'].str.replace('kleinwagen','small car').str.replace('kombi', 'combi').str.replace('cabrio', 'convertible').str.replace('andere', 'other')


In [43]:
df.vehicle_type.value_counts()

limousine      12546
small car      10439
combi           8896
bus             4022
convertible     3004
coupe           2453
suv             1961
other            386
Name: vehicle_type, dtype: int64

In [44]:
df['gearbox'].value_counts()

manuell      34453
automatik     9823
Name: gearbox, dtype: int64

In [45]:
df['gearbox'] = df['gearbox'].str.replace('manuell','manual').str.replace('automatik','automatic')
df['gearbox'].value_counts()

manual       34453
automatic     9823
Name: gearbox, dtype: int64

In [46]:
df['fuel_type'].unique()

array(['lpg', 'benzin', 'diesel', nan, 'cng', 'hybrid', 'elektro',
       'andere'], dtype=object)

In [47]:
df['fuel_type'] = df['fuel_type'].str.replace('benzin','petrol').str.replace('elektro','electric').str.replace('andere','others')
df.fuel_type.value_counts()

petrol      28292
diesel      14008
lpg           648
cng            70
hybrid         37
electric       18
others         14
Name: fuel_type, dtype: int64

In [48]:
df['unrepaired_damage'].unique()

array(['nein', nan, 'ja'], dtype=object)

In [49]:
df.unrepaired_damage.value_counts(dropna=False)

nein    33751
NaN      8035
ja       4434
Name: unrepaired_damage, dtype: int64

In [50]:
df['unrepaired_damage'].dropna(inplace=True)

In [51]:
df['unrepaired_damage'] = df['unrepaired_damage'].str.replace('ja','yes').str.replace('nein','no')
df.unrepaired_damage.value_counts()

no     33751
yes     4434
Name: unrepaired_damage, dtype: int64

Let's try to find the most common brand/model combination for the top brands:

In [52]:
brand_model = {}

for brand in top_brands:
    by_brand = df[df['brand'] == brand]
    top_model = by_brand['model'].describe()['top']
    brand_model[brand] = top_model
    
pprint.pprint(brand_model)

{'audi': 'a4',
 'bmw': '3er',
 'ford': 'focus',
 'mercedes_benz': 'c_klasse',
 'opel': 'corsa',
 'volkswagen': 'golf'}


Now let's see if the average prices follow any pattern based on mileage.

In [53]:
ranges = df['odometer_km'].value_counts(bins=5).index
ranges

IntervalIndex([(121000.0, 150000.0], (63000.0, 92000.0], (34000.0, 63000.0], (4854.999, 34000.0], (92000.0, 121000.0]],
              closed='right',
              dtype='interval[float64]')

In [54]:
ranges_tuple = ranges.to_tuples()

km_price = {}
for i in ranges_tuple:
    by_km = df[df['odometer_km'].between(i[0], i[1])]
    price = by_km['price'].mean()
    km_price[i] = int(price)
    
pprint.pprint(km_price)

{(4854.999, 34000.0): 15763,
 (34000.0, 63000.0): 13758,
 (63000.0, 92000.0): 9592,
 (92000.0, 121000.0): 8204,
 (121000.0, 150000.0): 4145}


In the above two cells we did the following:

we split the 'odometer_km' values in five bins (ranges)
we calculated the mean price for each odometer bin
We can conclude that the lower the milage of the car, the higher the price.

# How much cheaper are cars with damage than their non-damaged counterparts?

In [55]:
damaged = df['unrepaired_damage'].unique()

damaged_price = {}
for d in damaged:
    by_damaged = df[df['unrepaired_damage'] == d]
    price = by_damaged['price'].mean()
    damaged_price[d] = price 
    
pprint.pprint(damaged_price)

{nan: nan, 'no': 7181.541406180558, 'yes': 2293.0665313486693}


On average, cars with unrepaired damage are 30% cheaper than undamaged cars.

# Conclusion¶
In this guided project we explored data collected from the German eBay. The data contained information on car sales listings. We cleaned up the data and looked for any corelation between the average price and mileage of the top brands.